Add distributed RDataFrame blog post (#615)
root-project/web: blog/_posts/2021-07-06-distributed-rdataframe.md

---
title: "RDataFrame is going distributed!"
layout: archive
author: Vincenzo Eduardo Padulano
---

> A recently introduced Python package enables distributing ROOT RDataFrame workloads to a set of remote resources.

Author: Nice thanks!

This feature is available in experimental phase since the latest ROOT 6.24 release, allowing users to write and run
their applications from within the same interface while steering the computations to an Apache Spark cluster.

> ## In detail

Reviewer: For this title, what about: `One programming model, many backends`

Author: Not sure. It's not the point of the section. I agree that "In detail" is generic, but I think it has to be because we're not describing only the programming model in this text. Still, I do not prefer the current title either so I am willing to change it if others also agree

> [RDataFrame](https://root.cern/doc/master/classROOT_1_1RDataFrame.html) is ROOT's high-level interface for data
> analysis and since its release in 6.14 has seen tons of usages in real world analysis use cases and other examples in
> the wild. Parallelism has always been a staple of its design with support for executing the event loop on all cores

Reviewer 2: ```suggestion
[RDataFrame](https://root.cern/doc/master/classROOT_1_1RDataFrame.html) is ROOT's high-level interface for data
analysis since ROOT v6.14. By now many real world analyses use is, and on top of that we see lots of non-analysis usage in
the wild. Parallelism has always been a staple of its design with support for executing the event loop on all cores
```

of a machine thanks to implicit multi-threading. Since ROOT 6.24, this aspect of RDataFrame has been enhanced further
with distributed computing capabilities, allowing users to run their analysis on multi-node computing clusters through
widely used frameworks such as [Apache Spark](https://spark.apache.org/), [Dask](https://dask.org/) or
[AWS Lambda](https://aws.amazon.com/lambda/). Here is an example usage of the tool (requires the Python `pyspark`
package):

```python
import ROOT
 
# Point RDataFrame calls to the Spark specific RDataFrame
RDataFrame = ROOT.RDF.Experimental.Distributed.Spark.RDataFrame
 
# It still accepts the same constructor arguments as traditional RDataFrame

> df = RDataFrame("mytree", "myfile.root")

Reviewer: Shouldnt' we at least say that this will run in a local Spark process by default, but it's possible to configure the RDataFrame instance to use a preexisting Spark cluster?

> # Continue the application with the traditional RDataFrame API

Reviewer 2: ```suggestion
# Continue with the traditional RDataFrame API
```

sum = df.Filter("x > 10").Sum("y")
h = df.Histo1D("x")
 
print(sum.GetValue())
h.Draw()
```

The main goal is to support distributed execution of any RDataFrame application. This has led to the creation of a 
**Python** package that connects the RDataFrame API (available in Python through PyROOT) and the APIs of
distributed computing frameworks, which are offered in Python in the vast majority of cases. Another key goal is to be

> able to support as many backends as possible, in order to avoid being tied to any particular infrastructure. This is

Reviewer 2: Users are tied because they will have one infra - that "tying" is a developer's pov. What about:
```suggestion
able to support as many backends as possible, in order to provide a solution for as many users as possible. This is
```

Author: This is slightly clashing with Enric's comment from above. As I wrote there, I can maybe "merge" the two comments and write:
`Another key goal is to offer a variety of backends, to provide a solution to a variety of use cases` ?

> able to support as many backends as possible, in order to avoid being tied to any particular infrastructure. This is

Reviewer: I don't know if `as many as possible` is what we want, I would say quality better than quantity :) Should we say `Another key goal is to offer a variety of backends, ...`

Author: Ok, I can maybe "merge" it with Axel's suggestion below and write
`Another key goal is to offer a variety of backends, to provide a solution to a variety of use cases` ?

achieved through a modular implementation that defines a generic task (representing the RDataFrame computation graph) to
be executed on data. The input dataset is logically split into many ranges of entries, which will be sent to the
distributed nodes for processing. Each range will then be paired to the generic task and submitted to the computing
framework via a specific backend implementation. An added benefit of using RDataFrame is that the distributed
tasks run C++ computations. This is made possible by PyROOT and cling. Currently the package offers support for running

> the application on a Spark cluster, but the package design will make it possible to add many more backends in the future.

Reviewer: This sounds a bit repetitive to me w.r.t. what was written above (it's the second time the three backends are mentioned). Perhaps remove this and include the info about what is there (Spark) and what will come (Dask, AWS) in the first paragraph where you mention them?

Author: I actually made the post explicitly repetitive :smile: I think it's not easy to grasp all the meaning of this passage in a single sentence, plus the first paragraph it's a sort of "introduction" and just serves to mention the three backends currently "under our grasp", while more info about them is described in the next paragraphs.

Author: I can of course try to avoid repetitions, let me come up with something and I'll get back to you

For example, backends for Dask and AWS Lambda have already been implemented and demonstrated at different conferences
[[1]](#dasksummit)[[2]](#hips21). They will be made available in future ROOT releases.

> Great attention is being put towards ensuring that running the RDataFrame computation graph scales well across
> multiple computing nodes and different backend implementations. This has been shown since the first stages of the

Reviewer 2: ```suggestion
Excellent scaling is paramount for this distributed RDataFrame implementation, to ensure you can run the RDataFrame computation graph scales efficiently across
multiple computing nodes and different backend implementations. This has been shown since the first stages of the
```

development of this package with a real use case analysis running on a Spark cluster [[3]](#europar19). More recently, a benchmark
based on CERN open data has shown promising scaling performance with both Spark and Dask [[4]](#pyhep21):

![Scaling of distributed RDataFrame running the dimuon analysis on 100 times the original data. Dask and Spark backends compared]({{ '/assets/images/distrdf_dimuon_scaling_dask_spark_pyhep2021.png' | relative_url }})

> All it takes to steer an existing RDataFrame application to distributed resources is to use the correct RDataFrame

Reviewer 2: ```suggestion
All it takes to promote an existing RDataFrame application to distributed resources is to use the correct RDataFrame
```
or (more neutral) "move"

instance and provide it with the arguments needed to connect to the remote cluster. Let's see an example of connection
to a Dask cluster:

```python
from dask.distributed import Client
import ROOT

# Point RDataFrame calls to the Dask specific RDataFrame
RDataFrame = ROOT.RDF.Experimental.Distributed.Dask.RDataFrame

# Create the Client object to connect to the Dask cluster
# See the Dask documentation for all the options available
client = Client("DASK_SCHEDULER_ADDRESS")

# It still accepts the same constructor arguments as traditional RDataFrame
# And supports some extra keyword arguments
df = RDataFrame("mytree", "myfile.root", npartitions = 8, daskclient = client)
```

In this example the `npartitions` parameter tells the RDataFrame into how many ranges of entries the input dataset should
be split. Each range will then correspond to a task on a node of the cluster. The `daskclient` parameter receives the
object needed to connect to the Dask scheduler. All the options are available in the
[Dask documentation](https://distributed.dask.org/en/latest/client.html). The equivalent object for the Spark framework

> is called [SparkContext](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.SparkContext.html) and in

Reviewer: I think the two code examples (Spark and Dask) should go one after the other. This way you can explain well what I already pointed out in one of my previous comments: you can choose not to provide any connection details for the scheduler (Spark example above) and then everything will happen locally, but you can also configure RDataFrame to use an existing cluster (and you show an example with Dask). It just seems to me that the programming model description and the examples are scattered for no reason.

Author: My way of thinking about the distance between the two snippets is to avoid seeing too much code, too close. I think it would be confusing. I do believe you have a good point about the "you can use defaults vs you can provide more connection details". I can still come with something that is closer to your suggestion and send some screenshots to let you see

general every backend will have its own way to connect to a cluster of nodes.

Once the correct `RDataFrame` object has been created, there is no need to modify any other part of the program.

## Conclusions

Distributed RDataFrame enables large-scale interactive data analysis with ROOT. This Python layer on top of RDataFrame
allows to steer C++ computations on a set of computing nodes and returning the final result directly to the user, so
that the entire analysis can be run within the same application. It is available as an experimental feature since ROOT

> 6.24 with the support for running on a Spark cluster, with more backends coming in future ROOT releases. Try it with our

Reviewer 2: ```suggestion
6.24 with the support for running on a Spark cluster, with more backends in the works: Dask is already available in our nightly builds. Try it with our
```

Author: Ok so while it is true that the Dask backend could be considered "done" in a sense, it's not in master so it's not really in our nightly builds. I'm not sure it's good to write it that way

Reviewer 2: "will soon be available in our nighly builds"?

---
title: "RDataFrame is going distributed!"
layout: archive
author: Vincenzo Eduardo Padulano
---

> 

Reviewer 3: please make the first paragraph a short, self-contained intro: it's what appears in the blog preview at https://root.cern

> [RDataFrame](https://root.cern/doc/master/classROOT_1_1RDataFrame.html) is ROOT's declarative interface for data

Reviewer 3: ```suggestion
[RDataFrame](https://root.cern/doc/master/classROOT_1_1RDataFrame.html) is ROOT's high-level interface for data
```

SQL is fully declarative, RDF adopts elements of declarative programming in the API (but you still have to write imperative code for each `Filter` and `Define`). I usually don't get hung up on the details but since we are here :smile:

analysis and since its release in 6.14 has seen tons of usages in real world analysis use cases and other examples in

> the wild. Parallelism has always been a staple of its design with the support of executing the event loop on all cores

Reviewer 3: ```suggestion
the wild. Parallelism has always been a staple of its design with support for executing the event loop on all cores
```

of a machine thanks to implicit multi-threading. Since ROOT 6.24, this aspect of RDataFrame has been enhanced further
with distributed computing capabilities, allowing users to run their analysis on multi-node computing clusters through
widely used frameworks such as [Apache Spark](https://spark.apache.org/), [Dask](https://dask.org/) or
[AWS Lambda](https://aws.amazon.com/lambda/). The feature is currently in an experimental phase, try it out with this
simple example (requires the python `pyspark` package):

```python
import ROOT
 
# Point RDataFrame calls to the Spark specific RDataFrame
RDataFrame = ROOT.RDF.Experimental.Distributed.Spark.RDataFrame
 
# It still accepts the same constructor arguments as traditional RDataFrame
df = RDataFrame("mytree", "myfile.root")
 
# Continue the application with the traditional RDataFrame API
sum = df.Filter("x > 10").Sum("y")
h = df.Histo1D("x")
 
print(sum.GetValue())
h.Draw()
```

## Goals and implementation

> The main goal is to support distributed execution of any RDataFrame application. This has lead to the creation of a 

Reviewer 3: ```suggestion
The main goal is to support distributed execution of any RDataFrame application. This has led to the creation of a 
```

Author: I wanted to pour in some heavy metal :laughing:

> **Python** package that acts as a glue between the RDataFrame API (available in Python through PyROOT) and the APIs of

Reviewer 3: ```suggestion
**Python** package that connects the RDataFrame API (available in Python through PyROOT) and the APIs of
```

distributed computing frameworks, which are offered in Python in the vast majority of cases. Another key goal is to be
able to support as many backends as possible, in order to avoid being tied to any particular infrastructure. This is

> reached through a modular implementation, that defines a generic task (representing the RDataFrame computation graph) to

Reviewer 3: ```suggestion
achieved through a modular implementation that defines a generic task (representing the RDataFrame computation graph) to
```

be executed on data. The input dataset is logically split into many ranges of entries, which will be sent to the
distributed nodes for processing. Each range will then be paired to the generic task and submitted to the computing

> framework thanks to the particular backend implementation. An added benefit of using RDataFrame is that the distributed

Reviewer 3: ```suggestion
framework via a specific backend implementation. An added benefit of using RDataFrame is that the distributed
```

> tasks run C++ computations. This is made available thanks to PyROOT and cling.The current status of the package offers

Reviewer 3: ```suggestion
tasks run C++ computations. This is made possible by PyROOT and cling. Currently, the package offers
```

* AWS Lambda backend: developed thanks to an external contribution from AGH University and shown at [HPDC 2021](https://highperformanceserverless.github.io/program.html)

It will be possible to implement many more backends in the future thanks to this modular design.

As shown by the example snippet above, all it takes to steer an existing RDataFrame application to distributed
resources is to use the correct RDataFrame instance and provide it with the arguments needed to connect to the remote
cluster. Let's see an example of connection to a Dask cluster:

```python
from dask.distributed import Client
import ROOT

# Point RDataFrame calls to the Dask specific RDataFrame
RDataFrame = ROOT.RDF.Experimental.Distributed.Dask.RDataFrame

# Create the Client object to connect to the Dask cluster
# See the Dask documentation for all the options available
client = Client("DASK_SCHEDULER_ADDRESS")

# It still accepts the same constructor arguments as traditional RDataFrame
# And supports some extra keyword arguments
df = RDataFrame("mytree", "myfile.root", npartitions = 8, daskclient = client)
```

> In this example the `npartitions` parameter tells the RDataFrame in how many ranges of entries the input dataset should

Reviewer 3: ```suggestion
In this example the `npartitions` parameter tells the RDataFrame into how many ranges of entries the input dataset should
```
